### DISCUSSION TITLE
Looking for way to add correlation ids

### DISCUSSION THREAD

**Discussion by pupca** [RobinTail/express-zod-api#1306] [Q&A] [answered]
Hi,

I want to be able to add correlation ids to my app using express-zod-api.

My initial idea was to use `express-http-context` that would allow me to create a middleware such as

```
import { Request, Response, NextFunction } from 'express';
import { v4 } from 'uuid';
import * as httpContext from 'express-http-context';

export default async function requestId(req: Request, res: Response, next: NextFunction) {
  httpContext.ns.bindEmitter(req);
  httpContext.ns.bindEmitter(res);
  const id = v4();  
  httpContext.set('requestId', id);
  res.setHeader('request-id', id);
  console.log('request Id set is: ', httpContext.get('requestId'));
  return next();
}
```

This middleware would be added to my custom endpoint

```
export const jsonEndpointsFactory = new EndpointsFactory({
  resultHandler: defaultResultHandler,
  config,
})
  .addExpressMiddleware(cors({ credentials: true }))
  .addExpressMiddleware(requestId);
```

The problem is that the console output is

`request Id set is:  undefined`

The ultimate goal is to have a custom formatter in my winston that would send the request id to Loki.

One suggestion is that it is caused by https://stackoverflow.com/questions/55611335/node-js-express-unable-to-retrieve-value-from-http-context-for-post-and-put-re/58546037#58546037

But i am not able to change the order of middleware's here :( 

@RobinTail any idea what to do here, or how to achieve the ability to have a request id in my log classes? 

Thanks

**Comment by RobinTail** [comment]
Hello @pupca .

Before I get to your actual question, I noticed something I'd like to suggest.
Consider replacing 

```ts
.addExpressMiddleware(cors({ credentials: true }))
```

with an option in your config:

```ts
createConfig({
  cors: ({ defaultHeaders }) => ({ ...defaultHeaders, "Access-Control-Allow-Credentials": true })
})
```

It's acts the same, according to [documentation](https://www.npmjs.com/package/cors#configuring-cors) and [specification](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials), but this way you can reduce your dependencies and keep the control on CORS headers in one place.

Regarding `express-http-context` I need a bit more time to explore the problem.

**Comment by RobinTail** [comment]
> One suggestion is that it is caused by https://stackoverflow.com/questions/55611335/node-js-express-unable-to-retrieve-value-from-http-context-for-post-and-put-re/58546037#58546037
> But i am not able to change the order of middleware's here :(

They recommend to place body parser before, but it's already before every `.addExpressMiddleware()` you're using.

**Comment by pupca** [comment]
@RobinTail first of all, thank you very much for looking into that!

I have modified the cors settings, thank you!

Perhaps you can think of any other solution that would make correlation id available from within logger classes...

**Comment by RobinTail** [comment, marked as answer]
@pupca , here is what I found out.
This solution works fine, when you apply the `express-http-context` middleware directly to express app (not by using `.addExpressMiddleware`).
Therefore, consider going with custom express app approach and `attachRouting()` method, [described in the documentation](https://github.com/RobinTail/express-zod-api#connect-to-your-own-express-app), and applying the following lines to your own `app`:


```ts
import { v4 } from "uuid";
import * as httpContext from "express-http-context";

app.use(httpContext.middleware).use(({}, res, next) => {
    const requestId = v4();
    httpContext.set("requestId", requestId);
    console.log("the id", httpContext.get("requestId"));
    res.setHeader("x-request-id", requestId);
    next();
  });
```

I tested it and it works.

**Reply by pupca** [reply]
Thanks! 

that looks good. I think i got it working too.

Thanks again!
